### PR TITLE
[wip] hotfix/config-resource - Properly correct the ConfigResource object

### PR DESCRIPTION
**Pull request by TomHAnderson**
Apply PR changes from #25

### CODE REVIEW

File: src/ConfigResource.php [hunk 1]
```diff
@@ -39,8 +39,6 @@ class ConfigResource
 
     /**
      * @param array $config
-     * @param string $fileName
```
Comment (webimpress) [inline src/ConfigResource.php]: Why above line is removed?

File: src/ConfigResource.php [hunk 2]
```diff
@@ -35,32 +35,33 @@ class ConfigResource
     /**
      * @var ConfigWriter
      */
-    protected $writer;
+    protected $configWriter;
```
Comment (webimpress) [inline src/ConfigResource.php]: Why this property is renamed? It's `protected` and the class is not `final`, so it is 'minor BC break'.

Comment (TomHAnderson) [reply src/ConfigResource.php]: Thanks, that's a valid point.  I dislike abbreviations in code; call it what it is, not a nickname; but as you point out this isn't final that's a good catch.

**Comment by TomHAnderson** [comment]
This is ready for review/merge

**Comment by TomHAnderson** [comment]
@webimpress Thanks for taking an interest in this.  This repository, zf-configuration, is working IMHO now.  The most significant change besides correcting patchKey is patch() now uses a flag for ArrayUtils to preserve keys.  This is echoed in the last test.

I hope you'll review this at your convenience.

File: src/ConfigResource.php [hunk 3]
```diff
@@ -305,14 +356,16 @@ public function createNestedKeyValuePair(&$patchValues, $key, $value)
     protected function extractAndSet(array $keys, $value, &$array)
     {
         $key = array_shift($keys);
-        if ($keys) {
+        if (count($keys)) {
```
Comment (webimpress) [inline src/ConfigResource.php]: Why this change?

File: src/ConfigResource.php [hunk 4]
```diff
@@ -329,10 +382,12 @@ protected function deleteByKey(&$array, array $keys)
             return;
         }
 
-        if (! $keys) {
+        if (1 > count($keys)) {
```
Comment (webimpress) [inline src/ConfigResource.php]: ... and this?

File: src/ConfigResource.php [hunk 5]
```diff
@@ -184,17 +157,26 @@ public function fetch($tree = false)
      * @param  array $config
      * @return array
      */
-    public function replaceKey($keys, $value, array $config)
+    public function replaceKey($keys, $value, array $config = null)
```
Comment (webimpress) [inline src/ConfigResource.php]: Hm... BC Break, we can't set default value for a param, we will get warning/error. See: https://3v4l.org/imTjp

File: src/ConfigResource.php [hunk 6]
```diff
@@ -184,17 +157,26 @@ public function fetch($tree = false)
...
+            $config = $this->getConfig();
+        }
+
         if (! is_array($keys)) {
             $keys = explode('.', $keys);
         }
 
         $key = array_shift($keys);
 
+        $haveKeys = (bool) count($keys) > 0;
```
Comment (webimpress) [inline src/ConfigResource.php]: hm... I think we have here conversion twice, I mean it works like:
```
((bool) count($keys)) > 0
```
so first int -> bool and than comparison bool with int.

See:
https://3v4l.org/m0fnk
https://3v4l.org/TfFvI

I think we can just use `if ($keys)` (as we have in other places).

**Comment by webimpress** [comment]
@TomHAnderson I'm afraid this patch can't be applied as it is BC Break. See #29 and as I discussed with @weierophinney we can change it only in 2.0.0 release.
I know you provided changes to apigility-admin, but we have to think about other consumers of that library. They could also rely on that (broken) function and we can't change it in v1 :(

/cc @weierophinney

File: src/ConfigResource.php (hunk 4 above)
Comment (TomHAnderson) [reply src/ConfigResource.php]: I don't write code like that.  I think the reordering of protected functions to the end of the class moved these two together in the review.

File: src/ConfigResource.php (hunk 3 above)
Comment (TomHAnderson) [reply src/ConfigResource.php]: I think these are not the same rows as I don't recall changing this.

File: src/ConfigResource.php (hunk 5 above)
Comment (TomHAnderson) [reply src/ConfigResource.php]: This one is rather necessary else the whole config would need to be passed on the first call and this is a recursive function.  Passing the whole config is contrary to the class' job of managing a config and a file.  However as there are only a hand-ful of changes to replaceKey it would be possible to pass the config but I think that's not the best idea.

File: src/ConfigResource.php (hunk 6 above)
Comment (TomHAnderson) [reply src/ConfigResource.php]: That's fine with me.  Anything is better than the code that was there before.

**Comment by TomHAnderson** [comment]
I agree with you, @webimpress.  The library was so broken and was used even with that broken functionality that a 2.0 release to me, too, sounds like the only viable solution.  I do not think we should let this fester.  I want to see a new version released asap which we can use to correct zf-apigility-admin and zf-apigility-doctrine.

File: src/ConfigResource.php (hunk 4 above)
Comment (webimpress) [reply src/ConfigResource.php]: This is why we shouldn't do improvements with bug fixing because it's really confusing especially when you look on diff ...

**Comment by webimpress** [comment]
@TomHAnderson So what's wrong with #25? I think it should be applied for 2.0.
I'm a bit confused with your changes, because this PR contains bug fix and other improvements, so it's hard to review real changes in the functionality.

**Comment by TomHAnderson** [comment]
As you can see I started this work by merging back in #25 so it sounds like you're asking me from #25 what was each change in this PR made for.  Is that what you're asking?

The first commit after #25 major change was https://github.com/zfcampus/zf-configuration/pull/30/commits/9266764eb83217bd5d6b283f126b168d29f01785#diff-e813dca6ad9b65fb412e84d53d3d39a1R166 which exploded the key if it was an array.  My only lead going into this was a comment like you can't patch an array with array keys and that was because that person didn't realize the first array was a delimited list of the location in the tree so the code linked above was added.

In order for me to develop I needed to better understand what was taking place and that's why my first step in a refactor is breaking the parts up; you call this improvements.

Are you asking for a minimalist patch/PR?

**Comment by weierophinney** [comment]
Closing due to inactivity. My previous comment notes what needs to be done to get this accepted, should anyone want to pick this up.
